src/backend/common/CommandBufferStateTracker.cpp

```diff
@@ -0,0 +1,455 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "CommandBufferStateTracker.h"
+
+#include "Forward.h"
+#include "BindGroup.h"
+#include "BindGroupLayout.h"
+#include "BitSetIterator.h"
+#include "Buffer.h"
+#include "Framebuffer.h"
+#include "InputState.h"
+#include "Pipeline.h"
+#include "PipelineLayout.h"
+#include "RenderPass.h"
+#include "Texture.h"
+
+namespace backend {
+    CommandBufferStateTracker::CommandBufferStateTracker(CommandBufferBuilder* builder)
+        : builder(builder) {
+    }
+
+    bool CommandBufferStateTracker::HaveRenderPass() const {
+        return currentRenderPass != nullptr;
+    }
+
+    bool CommandBufferStateTracker::ValidateCanCopy() const {
+        if (currentRenderPass) {
+            builder->HandleError("Copy cannot occur during a render pass");
+            return false;
+        }
+        return true;
+    }
+
+    bool CommandBufferStateTracker::ValidateCanUseBufferAs(BufferBase* buffer, nxt::BufferUsageBit usage) const {
+        if (!BufferHasGuaranteedUsageBit(buffer, usage)) {
+            builder->HandleError("Buffer is not in the necessary usage");
+            return false;
+        }
+        return true;
+    }
+
+    bool CommandBufferStateTracker::ValidateCanUseTextureAs(TextureBase* texture, nxt::TextureUsageBit usage) const {
+        if (!TextureHasGuaranteedUsageBit(texture, usage)) {
+            builder->HandleError("Texture is not in the necessary usage");
+            return false;
+        }
+        return true;
+    }
+
+    bool CommandBufferStateTracker::ValidateCanDispatch() {
+        constexpr ValidationAspects requiredAspects =
+            1 << VALIDATION_ASPECT_COMPUTE_PIPELINE |
+            1 << VALIDATION_ASPECT_BIND_GROUPS;
+        if ((requiredAspects & ~aspects).none()) {
+            // Fast return-true path if everything is good
+            return true;
+        }
+
+        if (!aspects[VALIDATION_ASPECT_COMPUTE_PIPELINE]) {
+            builder->HandleError("No active compute pipeline");
+            return false;
+        }
+        // Compute the lazily computed aspects
+        if (!HaveAspectBindGroups()) {
+            builder->HandleError("Some bind groups are not set");
+            return false;
+        }
+        return true;
+    }
+
+    bool CommandBufferStateTracker::ValidateCanDrawArrays() {
+        // TODO(kainino@chromium.org): Check for a current render pass
+        constexpr ValidationAspects requiredAspects =
+            1 << VALIDATION_ASPECT_RENDER_PIPELINE |
+            1 << VALIDATION_ASPECT_BIND_GROUPS;
+        if ((requiredAspects & ~aspects).none()) {
+            // Fast return-true path if everything is good
+            return true;
+        }
+
+        return ValidateCanDraw();
```

Comment:
ValidateCanDraw is a bit confusing because it is so similar to ValidateCanDrawArrays, maybe rename it, for example with RevalidateCanDraw to mean it isn't lazy